packages/protocol/contracts/bridge/Bridge.sol

```diff
@@ -244,15 +243,26 @@ contract Bridge is EssentialContract, IBridge {
 
         emit MessageExecuted(msgHash);
 
-        refundAmount += _message.fee;
+        if (_message.fee != 0) {
+            refundAmount += _message.fee;
 
-        if (msg.sender != _message.destOwner) {
-            uint256 gasUsed = _calcGasToCharge(gas, GAS_OVERHEAD, _proof.length, numCacheOps);
-            uint256 fee = _calcFee(_message.fee, _message.gasLimit, gasUsed);
-            refundAmount -= fee;
-            msg.sender.sendEtherAndVerify(fee, _SEND_ETHER_GAS_LIMIT);
+            if (msg.sender != _message.destOwner && _message.gasLimit != 0) {
+                unchecked {
+                    uint256 gasUsed = gas - gasleft();
+                    emit GasChargedByRelayer(gasUsed, GAS_OVERHEAD);
 
-            emit GasChargedByRelayer(gasUsed);
+                    uint256 refund = numCacheOps * _GAS_REFUND_PER_CACHE_OPERATION;
+                    gasUsed = (GAS_OVERHEAD + gasUsed).max(refund) - refund;
```

Comment:
I think here :
`gasUsed = (GAS_OVERHEAD + gasUsed).max(refund) - refund;`

`refund` is always smaller than `(GAS_OVERHEAD+gasUsed)` since `gasUsed` already captures the amount of gas consumed with caching.

So maybe just this:
`gasUsed = (GAS_OVERHEAD + gasUsed) - refund;`

Reply:
The relayer could choose to cache a value that was already cached I think, in which case the SSTORE would be much cheaper than 20k (though the merkle proof verification is still extremely likely to be > 20k gas). 

But good to have the check there because it cannot be ruled out I think.